aspnetcore/fundamentals/app-state.md

```diff
@@ -1,236 +1,191 @@
 
-When the entry for the object exists in `Session`, it is retrieved as a `byte[]` type, and then deserialized using a `MemoryStream` and a `BinaryFormatter`, as shown above. If the object isn't in `Session`, the method returns a new instance of the `RequestEntryCollection`.
+In the sample below, [Middleware](middleware.md) adds `isVerified` to the `Items` collection:
 
-In the browser, clicking the Establish session hyperlink makes a request to the path "/session", and returns this result:
+```c#
 
-![Web application stating: Counting: You have made 1 requests to this application.](app-state/_static/session-established.png)
+   app.Use(async (context, next) =>
+   {
+       // perform some verification
+       context.Items["isVerified"] = true;
+       await next.Invoke();
+   });
+   ```
 
-Refreshing the page results in the count incrementing; returning to the root of the site (after making a few more requests) results in this display, summarizing all of the requests that were made during the current session:
+Later in the pipeline, another middleware could access it:
 
-![Web application after making several requests stating that session path was requested four times, the index path was requested four times, and the site has been visited eight times.](app-state/_static/session-established-with-request-counts.png)
+```c#
 
-Establishing the session is done in the middleware that handles requests to "/session":
+   app.Run(async (context) =>
+   {
+       await context.Response.WriteAsync("Verified request? " + context.Items["isVerified"]);
+   });
+   ```
 
-[!code-csharp[Main](../fundamentals/app-state/sample/src/AppState/Startup.cs?highlight=2,8,9,10,11,12,13,14&range=56-75)]
+Note: Since keys into `Items` are simple strings, if you are developing middleware that needs to work across many applications, you may wish to prefix your keys with a unique identifier to avoid key collisions (for example, "MyComponent.isVerified" instead of "isVerified").
 
-Requests to this path will get or create a `RequestEntryCollection`, will add the current path to it, and then will store it in session using the helper method `SaveEntries`, shown below:
+<a name=appstate-errors></a>
 
-[!code-csharp[Main](../fundamentals/app-state/sample/src/AppState/Startup.cs?highlight=6&range=126-132)]
+### Common errors when working with session
 
-`SaveEntries` demonstrates how to serialize a custom object into a `byte[]` for storage in `Session` using a `MemoryStream` and a `BinaryFormatter`.
+* "Unable to resolve service for type 'Microsoft.Extensions.Caching.Distributed.IDistributedCache' while attempting to activate 'Microsoft.AspNetCore.Session.DistributedSessionStore'."
 
-The sample includes one more piece of middleware worth mentioning, which is mapped to the "/untracked" path. You can see its configuration here:
+   Commonly caused by not configuring at least one `IDistributedCache` implementation.
```

Comment:
Add or link to recommended solutions. People like being given the code that solves their problem. InMemoryDistributedCache, SQL, Redis, etc..